pkg/controllers/cluster/taint_manager.go

```diff
@@ -0,0 +1,360 @@
+	return t, exist
+}
+
+func (n *evictionStartTimeMap) set(name string, time time.Time) {
+	n.Lock()
+	defer n.Unlock()
+	n.evictionStartTimes[name] = time
+}
+
+func (n *evictionStartTimeMap) delete(name string) {
+	n.Lock()
+	defer n.Unlock()
+	delete(n.evictionStartTimes, name)
+}
+
+// Reconcile performs a full reconciliation for the object referred to by the Request.
+// The Controller will requeue the Request to be processed again if an error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (tc *NoExecuteTaintManager) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	klog.V(4).Infof("Reconciling cluster %s for taint manager", req.NamespacedName.Name)
+
+	cluster := &clusterv1alpha1.Cluster{}
+	if err := tc.Client.Get(context.TODO(), req.NamespacedName, cluster); err != nil {
+		// The resource may no longer exist, in which case we stop processing.
+		if apierrors.IsNotFound(err) {
+			return controllerruntime.Result{}, nil
+		}
+		return controllerruntime.Result{Requeue: true}, err
+	}
+
+	// Check whether the target cluster has no execute taints.
+	if !helper.HasNoExecuteTaints(cluster.Spec.Taints) {
+		return controllerruntime.Result{}, nil
+	}
+
+	return tc.syncCluster(ctx, cluster)
+}
+
+func (tc *NoExecuteTaintManager) syncCluster(ctx context.Context, cluster *clusterv1alpha1.Cluster) (reconcile.Result, error) {
+	// List all ResourceBindings which are assigned to this cluster.
+	rbList := &workv1alpha2.ResourceBindingList{}
+	if err := tc.List(ctx, rbList, client.MatchingFieldsSelector{
+		Selector: fields.OneTermEqualSelector(rbClusterKeyIndex, cluster.Name),
+	}); err != nil {
+		klog.ErrorS(err, "Failed to list ResourceBindings", "cluster", cluster.Name)
+		return controllerruntime.Result{Requeue: true}, err
+	}
+	for i := range rbList.Items {
+		key, err := keys.FederatedKeyFunc(cluster.Name, &rbList.Items[i])
+		if err != nil {
+			klog.Warningf("Failed to generate key for obj: %s", rbList.Items[i].GetObjectKind().GroupVersionKind())
+			continue
+		}
+		tc.bindingEvictionWorker.Add(key)
+	}
+
+	// List all ClusterResourceBindings which are assigned to this cluster.
+	crbList := &workv1alpha2.ClusterResourceBindingList{}
+	if err := tc.List(ctx, crbList, client.MatchingFieldsSelector{
+		Selector: fields.OneTermEqualSelector(crbClusterKeyIndex, cluster.Name),
+	}); err != nil {
+		klog.ErrorS(err, "Failed to list ClusterResourceBindings", "cluster", cluster.Name)
+		return controllerruntime.Result{Requeue: true}, err
+	}
+	for i := range crbList.Items {
+		key, err := keys.FederatedKeyFunc(cluster.Name, &crbList.Items[i])
+		if err != nil {
+			klog.Warningf("Failed to generate key for obj: %s", crbList.Items[i].GetObjectKind().GroupVersionKind())
+			continue
+		}
+		tc.clusterBindingEvictionWorker.Add(key)
+	}
+	return controllerruntime.Result{RequeueAfter: tc.ClusterTaintEvictionRetryFrequency}, nil
```

Comment:
Well, the logic of taint manager is to list cluster which has NoExecute taint periodically and add all respective RBs and CRBs into eviction worker. So if something happens to a RB, e.g. it has an NoExecute toleration before but the toleration changes so that it should be evicted, we could not receive an event. Thus, we check the cluster periodically, it's a litlle bit like what `cluster_status_controller` does.

Reply:
got it. thanks.